ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -418,26 +436,25 @@ abstract class EventStorageBackendTemplate(
       val witnessesWhereClause =
         (wildcardPartiesClause ::: filterPartiesClauses).mkComposite("(", " or ", ")")
 
-      def selectFrom(table: String, selectColumns: String) = cSQL"""
+      def selectFrom(partition: EventPartition[T]) = SQL"""
         SELECT
-          #$selectColumns, #$witnessesColumn as event_witnesses, command_id
+          #${partition.selectColumns}, #$witnessesColumn as event_witnesses, command_id
         FROM
-          #$table $joinClause
+          #${partition.tableName} $joinClause
         WHERE
           $additionalAndClause
           $witnessesWhereClause
-      """
-
-      val selectClause = partitions
-        .map(p => selectFrom(p._1, p._2))
-        .mkComposite("", " UNION ALL", "")
-
-      SQL"""
-        $selectClause
         ORDER BY event_sequential_id
-        ${QueryStrategy.limitClause(limit)}"""
+        ${QueryStrategy.limitClause(limit)}
+      """
         .withFetchSize(fetchSizeHint)
-        .asVectorOf(rowParser(internedAllParties))(connection)
+        .asVectorOf(partition.parser(internedAllParties))(connection)
+
+      // TODO: we are merging multiple sorted Vectors and then only taking a few elements, this could be done more efficiently.
+      partitions
+        .flatMap(selectFrom)
+        .sorted(ordering)
+        .take(limit.getOrElse(Int.MaxValue))
```

Comment:
doesn't it mean that we might take upto 3 times more data then what we return?
can't we just return all, since we already parsed ordered and prepared all?